### PR TITLE
Add initial rough version of port based rules

### PR DESCRIPTION
**Pull request by conectado**
Addresses #810 

### Refactor `fz_wall` nft module to support port ranges:
  * Instead of having rules matching source_set <-> dest_set we have:
    * Rule matching source_set(device) to jump to a custom chain of the corresponding user
    * Each user chain matches either destination/port-type/port-range(This can only match layer4 headers) or only destination to still work with layer 3 protocols
 * Some additional refactor for cleanliness 
### Refactor `fz_http` to support port ranges:
  * Add `Int4Range` `Ecto.Type` to represent ranges
    * Used `"start - stop"` as representation otherwise with types like `[start, stop]` I had errors while sending the changeset through the socket (Had this error: `lists in Phoenix.HTML templates only support iodata, and not chardata. Integers may only represent bytes. It's likely you meant to pass a string with double quotes instead of a char list with single quotes.`) we can maybe revisit this later
  * Adds `port_range` and `port_type` fields to the rules' schema
    * Updated constraint to require either both as `NULL` or neither 
    * Require valid port ranges 
    * Updated the overlaps constraints to take these new fields into account
#### When no port-type is selected user can't put a port range

image

#### When a port-type is selected user can 
     
image

#### Updated table to show the port range
      
image

#### And single port numbers are supported

image

### CODE REVIEW

File: apps/fz_http/lib/fz_http/int4range.ex [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      {:ok, range} -> cast(range)
+      err -> err
+    end
+  end
+
+  def cast([num, num]) when is_number(num) do
+    {:ok, Integer.to_string(num)}
+  end
+
+  def cast([lower, upper]) do
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/int4range.ex]: Could use a guard:

```elixir
def cast([lower, upper]) when upper >= lower, do: {:ok, "#{lower} - #{upper}"}
def cast([_, _]), do: {:error, message: "Range Error: Lower bound higher than upper bound"}
```

File: apps/fz_http/lib/fz_http/int4range.ex [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      |> Enum.map(&Integer.parse/1)
+
+    case res do
+      [{lower, _}, {upper, _}] -> {:ok, [lower, upper]}
+      [{num, _}] -> {:ok, [num, num]}
+      _ -> {:error, message: "Range Error: Bad format"}
+    end
+  end
+
+  defp to_num(b), do: if(b, do: 1, else: 0)
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/int4range.ex]: Could use a guard here:

```elixir
defp to_num(b) when b, do: 1
defp to_num(_b), do: 0
```

File: apps/fz_http/lib/fz_http/int4range.ex (hunk 1 above)
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/int4range.ex:22]: Could use a guard here:

```elixir
def cast([lower, upper]) when upper >= lower, do: {:ok, "#{lower} - #{upper}"}
```

File: apps/fz_http/lib/fz_http/int4range.ex [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  def cast([lower, upper]) do
+    if upper >= lower do
+      {:ok, "#{lower} - #{upper}"}
+    else
+      {:error, message: "Range Error: Lower bound higher than upper bound"}
+    end
+  end
+
+  def cast(_), do: :error
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/int4range.ex]: Is this expected to happen or you think we can just let the function call fail?

File: apps/fz_http/lib/fz_http/int4range.ex [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  def cast([num, num]) when is_number(num) do
+    {:ok, Integer.to_string(num)}
+  end
+
+  def cast([lower, upper]) do
+    if upper >= lower do
+      {:ok, "#{lower} - #{upper}"}
+    else
+      {:error, message: "Range Error: Lower bound higher than upper bound"}
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/int4range.ex]: If this is expected to happen should it be a validation instead?

File: apps/fz_http/lib/fz_http/rules.ex [hunk 5]
```diff
@@ -19,14 +20,27 @@ defmodule FzHttp.Rules do
 
   def as_settings do
     Repo.all(
-      from r in Rule, select: %{destination: r.destination, user_id: r.user_id, action: r.action}
+      from r in Rule,
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/rules.ex]: Would it be cleaner just to return rules as structs?

```elixir
Repo.all(Rule)
|> Enum.map(&setting_projection/1)
|> MapSet.new()
```

File: apps/fz_http/lib/fz_http/rules.ex [hunk 6]
```diff
@@ -19,14 +20,27 @@ defmodule FzHttp.Rules do
...
 
   def setting_projection(rule) do
-    %{destination: decode(rule.destination), user_id: rule.user_id, action: rule.action}
+    %{
+      destination: decode(rule.destination),
+      user_id: rule.user_id,
+      action: rule.action,
+      port_range: rule.port_range,
+      port_type: rule.port_type
+    }
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/rules.ex]: Maybe could be shortened?

```suggestion
rule
|> Map.take([:user_id, :action, :port_range, :port_type])
|> Map.merge(%{destination: decode(rule.destination)})
```

File: apps/fz_http/lib/fz_http/rules/rule.ex [hunk 7]
```diff
@@ -7,11 +7,15 @@ defmodule FzHttp.Rules.Rule do
   import Ecto.Changeset
 
   @exclusion_msg "Destination overlaps with an existing rule"
+  @port_range_msg "Port is not within valid range"
+  @port_type_msg "Please specify a port-range for the given port type"
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http/rules/rule.ex:11]: `port-range` -> `port range` :-)

File: apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.ex [hunk 8]
```diff
@@ -79,4 +91,12 @@ defmodule FzHttpWeb.RuleLive.RuleListComponent do
   defp user_options(users) do
     Enum.map(users, fn {id, email} -> {email, id} end)
   end
+
+  defp port_type_options do
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.ex:90]: Make this a module attribute?

```elixir
@port_type_opts %{TCP: :tcp, UDP: :udp}
```

**Review by jamilbk — CHANGES_REQUESTED** [review]
Suggestions on cleaning things, but looks good overall. Could you post up a screenshot of the new port range UI in this PR?

File: apps/fz_http/lib/fz_http/int4range.ex (hunk 4 above)
Comment (conectado) [reply apps/fz_http/lib/fz_http/int4range.ex]: I tried adding it as a validation check but the moment you add an int4range to the DB where this happen it fails with an exception before the validation runs.

File: apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.ex (hunk 8 above)
Comment (conectado) [reply apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.ex:90]: I tried to use a module attribute but it seems that the only way I could then access the attribute from the `heex` is to assign it to the socket, the function seems more tidy than doing that.

**Comment by conectado** [comment]
@jamilbk I added the UI screenshots in the PR description. Do those work or should I add some more?

File: apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.ex (hunk 8 above)
Comment (jamilbk) [reply apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.ex:90]: Kk

**Review by jamilbk — APPROVED** [review]
LGTM. May need some design refresh but can save that for another issue.

**Pull request merged**